src/end_entity.rs

```diff
@@ -74,6 +75,37 @@ impl<'a> EndEntityCert<'a> {
         &self.inner
     }
 
+    /// Verifies that the end-entity certificate is valid for use against the
+    /// specified EKU (Extended Key Use).
+    ///
+    /// `supported_sig_algs` is the list of signature algorithms that are
+    /// trusted for use in certificate signatures; the end-entity certificate's
+    /// public key is not validated against this list. `trust_anchors` is the
+    /// list of root CAs to trust. `intermediate_certs` is the sequence of
+    /// intermediate certificates that the server sent in the TLS handshake.
+    /// `time` is the time for which the validation is effective (usually the
+    /// current time).
+    pub fn verify_is_valid_cert_with_eku(
+        &self,
+        required_eku: KeyPurposeId,
```

Comment:
There's a little bit of nuance here that I think should be reflected in the parameter name and doc string.

The `KeyPurposeId` argument here is fed down through to [`check_eku`](https://github.com/rustls/webpki/blob/4a39e2b67d4cddf58b0ea16dd821a04ee2240058/src/verify_cert.rs#L371-L374) in `verify_cert.rs` for the actual validation. That function will return `Ok(())` if the certificate has an EKU specified that includes the required EKU, **or has no EKU specified at all**.

That's the reason the `verify_cert::build_chain` argument is named `required_eku_if_present` instead of `required_eku` and is something I think we should match in this function.

Is that existing behaviour acceptable for the use case you had in mind or would you be expecting `verify_is_valid_cert_with_eku` to return an error if the certificate had no EKU at all?

Reply:
Right, that's a good point. I noticed that behaviour of `check_eku`, not sure why I didn't think to call the parameter `required_eku_if_present` like it is there.

ISO 18013-5 requires that the EKU is always present. I thought it might be best to keep this PR small and not modify this aspect of this crate, so in my calling code I now use `x509_parser` (which I already used anyway) to actually enforce the EKU being present.

However, if you agree I would prefer to add an additional boolean called `enforce_eku_presence` or something, pass that along to `check_eku`, and then actually enforce the EKU being pesent if that parameter is set to true. What do you think about that? Alternatively, I'll rename `required_eku` to `required_eku_if_present` instead.

Reply:
> However, if you agree I would prefer to add an additional boolean called enforce_eku_presence or something, pass that along to check_eku, and then actually enforce the EKU being pesent if that parameter is set to true. What do you think about that?

I think exposing some flexibility for requiring an EKU full-stop seems reasonable to me.

Instead of adding an additional bool parameter, what about changing the `ChainOpts.required_eku_if_present` field from `KeyPurposeId` to an exported enum that has two variants:

```rust
pub enum ExtendedKeyUsage {
    Required(KeyPurposeId),
    RequiredIfPresent(KeyPurposeId),
}
```

The existing code in `verify_is_valid_tls_client_cert` and `verify_is_valid_tls_server_cert` could be updated to build `ChainOptions` with `RequiredIfPresent(verify_cert::EKU_CLIENT_AUTH)` and `RequiredIfPresent(verify_cert::EKU_SERVER_AUTH)`.

Your new `verify_is_valid_cert_with_eku` function could take an `ExtendedKeyUsage` instead of `KeyPurposeId` and pass it straight through to the `ChainOptions` field, letting the caller decide how to enforce the EKU.

Does that seem workable to you? It's possible other maintainers will have their own opinions but I think that would be a nicer representation than an additional `bool` arg.

Reply:
That looks good, I'll do that!